File: etc3460_toolkit.lua
platform.apilevel = '2.0'

local screen = "menu"
local choice = 1
local inputIndex = 1
local inputText = ""
local values = {}
local output = {}

local items = {
  {name="Coeff test", prompts={"est","se","h0","crit","type","mode"}, fn="ctest"},
  {name="Risk split", prompts={"R2","mode"}, fn="risk"},
  {name="GMV weights", prompts={"var1","var2","cov12","mode"}, fn="gmv2"},
  {name="Two-asset portfolio", prompts={"w1","ret1","ret2","var1","var2","cov12","mode"}, fn="port2"},
  {name="RSS from RSE", prompts={"RSE","df","mode"}, fn="rssp"},
  {name="Joint chi test", prompts={"RSSr","RSSu","dfu","crit","mode"}, fn="jchi"},
  {name="Joint F test", prompts={"RSSr","RSSu","q","dfu","crit","mode"}, fn="jf"},
  {name="ADF decision", prompts={"ADF stat","crit","mode"}, fn="adf"},
  {name="AR1 moments", prompts={"c","phi","innov var","mode"}, fn="ar1m"},
  {name="AR1 forecast", prompts={"c","phi","last","mode"}, fn="ar1f"},
  {name="Mean adj AR forecast", prompts={"mu0","phi","last","mode"}, fn="ar1fm"},
  {name="GARCH forecast", prompts={"omega","alpha","beta","resid","var","mode"}, fn="gar1"},
  {name="GARCH persistence", prompts={"alpha","beta","mode"}, fn="gper"},
  {name="Prediction interval", prompts={"mu","var","z","mode"}, fn="pi1"},
  {name="VaR percent returns", prompts={"value","mu pct","var","z","mode"}, fn="varp"},
  {name="VaR decimal returns", prompts={"value","mu dec","var","z","mode"}, fn="vard"},
  {name="ARCH LM", prompts={"n/T","R2","q","crit","mode"}, fn="arch"},
  {name="EMH residual test", prompts={"n/T","R2","p","crit","mode"}, fn="emh"},
  {name="Jarque-Bera", prompts={"n","skew","kurt","crit","mode"}, fn="jb"},
  {name="GJR curvature", prompts={"alpha","lambda","mode"}, fn="gjr"},
  {name="AIC comparison", prompts={"AIC1","AIC2","mode"}, fn="aic"}
}

local function rnd(x)
  if not x then return "nil" end
  return string.format("%.6g", x)
end

local function add(s)
  output[#output+1] = tostring(s)
end

local function mode(v)
  return tonumber(v or 0) == 1
end

local function ctest(v)
  local est,se,h0,crit,typ,m = v[1],v[2],v[3],v[4],v[5],mode(v[6])
  local t = (est-h0)/se
  local at = math.abs(t)
  if m then add("t=(est-h0)/se"); add("est="..rnd(est).." h0="..rnd(h0)); add("se="..rnd(se)) end
  add("t="..rnd(t))
  if m then add("|t|="..rnd(at).." crit="..rnd(crit)) end
  if at > crit then
    add("Reject H0")
    if typ==2 then add("Beta differs from h0"); if est>1 then add("Beta>1: aggressive") elseif est<1 then add("Beta<1: defensive") end
    elseif typ==3 then add("Alpha differs from 0"); add("Abnormal return")
    elseif typ==4 then add("SMB significant"); if est>0 then add("Positive: small tilt") else add("Negative: big tilt") end
    elseif typ==5 then add("HML significant"); if est>0 then add("Positive: value tilt") else add("Negative: growth tilt") end
    elseif typ==6 then add("MOM significant"); if est>0 then add("Positive: winner tilt") else add("Negative: loser tilt") end
    elseif typ==7 then add("Leverage term matters")
    else add("Coeff differs from h0") end
  else
    add("Fail reject H0")
    if typ==4 then add("SMB not significant")
    elseif typ==5 then add("HML not significant")
    elseif typ==6 then add("MOM not significant")
    elseif typ==7 then add("No leverage evidence")
    else add("Insufficient evidence") end
  end
end

local function risk(v)
  local r2,m = v[1],mode(v[2])
  if m then add("systematic=R2"); add("idiosyn=1-R2") end
  add("systematic="..rnd(r2))
  add("idiosyn="..rnd(1-r2))
  add("sys%="..rnd(100*r2))
  add("idio%="..rnd(100*(1-r2)))
end

local function gmv2(v)
  local v1,v2,c12,m = v[1],v[2],v[3],mode(v[4])
  local num = v2-c12
  local den = v1+v2-2*c12
  local w1 = num/den
  if m then add("w1=(v2-cov)/den"); add("num="..rnd(num)); add("den="..rnd(den)) end
  add("w1="..rnd(w1))
  add("w2="..rnd(1-w1))
end

local function port2(v)
  local w1,r1,r2,v1,v2,c12,m = v[1],v[2],v[3],v[4],v[5],v[6],mode(v[7])
  local w2 = 1-w1
  local pret = w1*r1+w2*r2
  local pv = w1*w1*v1+w2*w2*v2+2*w1*w2*c12
  if m then add("w2=1-w1="..rnd(w2)); add("var=w1^2v1+w2^2v2+2w1w2cov") end
  add("port ret="..rnd(pret))
  add("port var="..rnd(pv))
  add("port sd="..rnd(math.sqrt(pv)))
end

local function rssp(v)
  local rse,df,m = v[1],v[2],mode(v[3])
  if m then add("RSS=RSE^2*df") end
  add("RSS="..rnd(rse*rse*df))
end

local function jchi(v)
  local rssr,rssu,dfu,crit,m = v[1],v[2],v[3],v[4],mode(v[5])
  local num = rssr-rssu
  local den = rssu/dfu
  local j = num/den
  if m then add("J=(RSSr-RSSu)/(RSSu/dfu)"); add("num="..rnd(num).." den="..rnd(den)) end
  add("J="..rnd(j).." crit="..rnd(crit))
  if j>crit then add("Reject restrictions"); add("Use larger model") else add("Fail reject"); add("Restricted ok") end
end

local function jf(v)
  local rssr,rssu,q,dfu,crit,m = v[1],v[2],v[3],v[4],v[5],mode(v[6])
  local num = (rssr-rssu)/q
  local den = rssu/dfu
  local f = num/den
  if m then add("F=((RSSr-RSSu)/q)/(RSSu/dfu)"); add("num="..rnd(num).." den="..rnd(den)) end
  add("F="..rnd(f).." crit="..rnd(crit))
  if f>crit then add("Reject restrictions"); add("Use larger model") else add("Fail reject"); add("Restricted ok") end
end

local function adf(v)
  local stat,crit,m = v[1],v[2],mode(v[3])
  if m then add("Reject if stat < crit"); add("stat="..rnd(stat).." crit="..rnd(crit)) end
  if stat<crit then add("Reject unit root"); add("Evidence stationary") else add("Fail reject unit root"); add("Treat non-stationary") end
end

local function ar1m(v)
  local c,phi,s2,m = v[1],v[2],v[3],mode(v[4])
  if m then add("mean=c/(1-phi)"); add("var=s2/(1-phi^2)") end
  add("mean="..rnd(c/(1-phi)))
  add("var="..rnd(s2/(1-phi*phi)))
  if math.abs(phi)<1 then add("abs(phi)<1: stat ok") else add("not stationary") end
end

local function ar1f(v)
  local c,phi,last,m = v[1],v[2],v[3],mode(v[4])
  if m then add("fc=c+phi*last") end
  add("forecast="..rnd(c+phi*last))
end

local function ar1fm(v)
  local mu0,phi,last,m = v[1],v[2],v[3],mode(v[4])
  if m then add("fc=mu0+phi*(last-mu0)") end
  add("forecast="..rnd(mu0+phi*(last-mu0)))
end

local function gar1(v)
  local w,a,b,u,var,m = v[1],v[2],v[3],v[4],v[5],mode(v[6])
  local shock = a*u*u
  local persist = b*var
  local vn = w+shock+persist
  if m then add("var=w+a*u^2+b*var"); add("shock="..rnd(shock)); add("persist="..rnd(persist)) end
  add("next var="..rnd(vn))
  add("next sd="..rnd(math.sqrt(vn)))
end

local function gper(v)
  local p,m = v[1]+v[2],mode(v[3])
  if m then add("persist=alpha+beta") end
  add("persist="..rnd(p))
  if p<1 then add("Finite var condition ok"); if p>.9 then add("Highly persistent") end else add("Condition fails") end
end

local function pi1(v)
  local mu,var,z,m = v[1],v[2],v[3],mode(v[4])
  local sd = math.sqrt(var)
  if m then add("PI=mu +/- z*sd"); add("sd="..rnd(sd)) end
  add("lower="..rnd(mu-z*sd))
  add("upper="..rnd(mu+z*sd))
end

local function varp(v)
  local value,mu,var,z,m = v[1],v[2],v[3],v[4],mode(v[5])
  local sd = math.sqrt(var)
  local q = mu-z*sd
  if m then add("Percent return VaR"); add("q=mu-z*sd"); add("VaR=abs(value*q/100)") end
  add("q percent="..rnd(q))
  add("VaR dollars="..rnd(math.abs(value*q/100)))
end

local function vard(v)
  local value,mu,var,z,m = v[1],v[2],v[3],v[4],mode(v[5])
  local sd = math.sqrt(var)
  local q = mu-z*sd
  if m then add("Decimal return VaR"); add("q=mu-z*sd"); add("VaR=abs(value*q)") end
  add("q decimal="..rnd(q))
  add("VaR dollars="..rnd(math.abs(value*q)))
end

local function arch(v)
  local stat = v[1]*v[2]
  local crit = v[4]
  if mode(v[5]) then add("ARCH LM=T*R2"); add("df="..rnd(v[3])) end
  add("LM="..rnd(stat).." crit="..rnd(crit))
  if stat>crit then add("Reject no ARCH"); add("Vol time-varying") else add("Fail reject"); add("No ARCH found") end
end

local function emh(v)
  local stat = v[1]*v[2]
  local crit = v[4]
  if mode(v[5]) then add("Residual AR test=T*R2"); add("df="..rnd(v[3])) end
  add("stat="..rnd(stat).." crit="..rnd(crit))
  if stat>crit then add("Reject no autocorr"); add("Against weak EMH") else add("Fail reject"); add("No EMH evidence against") end
end

local function jb(v)
  local n,sk,ku,crit,m = v[1],v[2],v[3],v[4],mode(v[5])
  local skp = sk*sk
  local kp = ((ku-3)^2)/4
  local j = n/6*(skp+kp)
  if m then add("JB=n/6*(S^2+(K-3)^2/4)"); add("S^2="..rnd(skp)); add("Kpart="..rnd(kp)) end
  add("JB="..rnd(j).." crit="..rnd(crit))
  if sk<0 then add("Skew<0: left tail") elseif sk>0 then add("Skew>0: right tail") end
  if ku>3 then add("K>3: fat tails") end
  if j>crit then add("Reject normality") else add("Fail reject normal") end
end

local function gjr(v)
  local a,l,m = v[1],v[2],mode(v[3])
  if m then add("good=alpha"); add("bad=alpha+lambda") end
  add("good="..rnd(a))
  add("bad="..rnd(a+l))
  if l>0 then add("Bad news bigger effect") elseif l==0 then add("Symmetric effect") else add("Bad news smaller effect") end
end

local function aic(v)
  local a1,a2,m = v[1],v[2],mode(v[3])
  if m then add("Lower AIC preferred"); add("AIC1="..rnd(a1).." AIC2="..rnd(a2)) end
  if a1<a2 then add("Choose model 1") elseif a2<a1 then add("Choose model 2") else add("AIC tie") end
end

local funcs = {ctest=ctest,risk=risk,gmv2=gmv2,port2=port2,rssp=rssp,jchi=jchi,jf=jf,adf=adf,ar1m=ar1m,ar1f=ar1f,ar1fm=ar1fm,gar1=gar1,gper=gper,pi1=pi1,varp=varp,vard=vard,arch=arch,emh=emh,jb=jb,gjr=gjr,aic=aic}

local function runCurrent()
  output = {}
  funcs[items[choice].fn](values)
  screen = "output"
  platform.window:invalidate()
end

local function startInput()
  screen = "input"
  inputIndex = 1
  inputText = ""
  values = {}
  output = {}
  platform.window:invalidate()
end

function on.paint(gc)
  gc:setFont("sansserif","r",10)
  local y = 4
  if screen == "menu" then
    gc:drawString("ETC3460 Toolkit",4,y,"top"); y=y+16
    gc:drawString("Up/Down choose, Enter select",4,y,"top"); y=y+14
    local first = math.max(1, math.min(choice-5, #items-10))
    for i=first, math.min(#items, first+10) do
      local prefix = "  "
      if i == choice then prefix = "> " end
      gc:drawString(prefix..i.." "..items[i].name,4,y,"top")
      y = y + 14
    end
  elseif screen == "input" then
    local it = items[choice]
    gc:drawString(it.name,4,y,"top"); y=y+16
    gc:drawString("Enter "..it.prompts[inputIndex]..":",4,y,"top"); y=y+16
    gc:drawString(inputText,4,y,"top"); y=y+18
    gc:drawString("Enter=next  Back=edit",4,y,"top"); y=y+14
    gc:drawString("Esc=menu. mode 0/1",4,y,"top")
  elseif screen == "output" then
    gc:drawString(items[choice].name.." result",4,y,"top"); y=y+16
    for i=1, math.min(#output, 12) do
      gc:drawString(output[i],4,y,"top")
      y = y + 14
    end
    gc:drawString("Enter/Esc: menu",4,y+4,"top")
  end
end

function on.arrowDown()
  if screen == "menu" then
    choice = choice + 1
    if choice > #items then choice = 1 end
    platform.window:invalidate()
  end
end

function on.arrowUp()
  if screen == "menu" then
    choice = choice - 1
    if choice < 1 then choice = #items end
    platform.window:invalidate()
  end
end

function on.enterKey()
  if screen == "menu" then
    startInput()
  elseif screen == "input" then
    local n = tonumber(inputText)
    if n == nil then
      inputText = ""
    else
      values[inputIndex] = n
      inputIndex = inputIndex + 1
      inputText = ""
      if inputIndex > #items[choice].prompts then runCurrent() end
    end
    platform.window:invalidate()
  elseif screen == "output" then
    screen = "menu"
    platform.window:invalidate()
  end
end

function on.escapeKey()
  screen = "menu"
  inputText = ""
  platform.window:invalidate()
end

function on.backspaceKey()
  if screen == "input" and string.len(inputText) > 0 then
    inputText = string.sub(inputText,1,string.len(inputText)-1)
    platform.window:invalidate()
  end
end

function on.deleteKey()
  on.backspaceKey()
end

function on.charIn(ch)
  if screen == "input" then
    if string.find("0123456789.-", ch, 1, true) then
      inputText = inputText .. ch
      platform.window:invalidate()
    end
  end
end

File: etc3460_toolkit.py
from math import sqrt


def fnum(prompt):
    while True:
        try:
            return float(input(prompt))
        except Exception:
            print("Enter a number.")


def inum(prompt):
    while True:
        try:
            return int(float(input(prompt)))
        except Exception:
            print("Enter an integer.")


def mode_prompt():
    m = inum("Mode 0 compact, 1 working: ")
    return 1 if m == 1 else 0


def decision(stat, crit):
    return stat > crit


def ctest(est, se, h0, crit, typ, mode=0):
    t = (est - h0) / se
    at = abs(t)
    if mode:
        print("t=(est-h0)/se")
        print("est=", est)
        print("h0=", h0)
        print("se=", se)
    print("t=", round(t, 6))
    if mode:
        print("|t|=", round(at, 6))
        print("crit=", crit)
    if at > crit:
        print("Reject H0")
        if typ == 1:
            print("Coeff differs from h0")
        elif typ == 2:
            print("Beta differs from h0")
            if est > 1:
                print("Beta>1: aggressive")
            elif est < 1:
                print("Beta<1: defensive")
        elif typ == 3:
            print("Alpha differs from 0")
            print("Abnormal return")
        elif typ == 4:
            print("SMB is significant")
            print("Positive: small tilt" if est > 0 else "Negative: big tilt")
        elif typ == 5:
            print("HML is significant")
            print("Positive: value tilt" if est > 0 else "Negative: growth tilt")
        elif typ == 6:
            print("MOM is significant")
            print("Positive: winner tilt" if est > 0 else "Negative: loser tilt")
        elif typ == 7:
            print("Leverage term matters")
    else:
        print("Fail reject H0")
        if typ == 1:
            print("Insufficient evidence")
        elif typ == 2:
            print("No clear beta diff")
        elif typ == 3:
            print("No abnormal return found")
        elif typ == 4:
            print("SMB not significant")
        elif typ == 5:
            print("HML not significant")
        elif typ == 6:
            print("MOM not significant")
        elif typ == 7:
            print("No leverage evidence")


def risk(r2, mode=0):
    sys = r2
    idio = 1 - r2
    if mode:
        print("systematic=R2")
        print("idiosyncratic=1-R2")
    print("systematic=", round(sys, 6))
    print("idiosyncratic=", round(idio, 6))
    print("sys%=", round(100 * sys, 4))
    print("idio%=", round(100 * idio, 4))


def gmv2(v1, v2, c12, mode=0):
    num = v2 - c12
    den = v1 + v2 - 2 * c12
    w1 = num / den
    w2 = 1 - w1
    if mode:
        print("w1=(v2-cov)/(v1+v2-2cov)")
        print("num=", round(num, 8))
        print("den=", round(den, 8))
    print("w1=", round(w1, 6))
    print("w2=", round(w2, 6))
    print("Risk min weights")


def port2(w1, r1, r2, v1, v2, c12, mode=0):
    w2 = 1 - w1
    pret = w1 * r1 + w2 * r2
    pv = w1 ** 2 * v1 + w2 ** 2 * v2 + 2 * w1 * w2 * c12
    ps = sqrt(pv)
    if mode:
        print("w2=1-w1=", round(w2, 6))
        print("ret=w1*r1+w2*r2")
        print("var=w1^2*v1+w2^2*v2+2*w1*w2*cov")
    print("port ret=", round(pret, 6))
    print("port var=", round(pv, 6))
    print("port sd=", round(ps, 6))


def rssp(rse, df, mode=0):
    rv = rse ** 2 * df
    if mode:
        print("RSS=RSE^2*df")
        print("RSE=", rse)
        print("df=", df)
    print("RSS=", round(rv, 6))


def jchi(rssr, rssu, dfu, crit, mode=0):
    num = rssr - rssu
    den = rssu / dfu
    j = num / den
    if mode:
        print("J=(RSSr-RSSu)/(RSSu/dfu)")
        print("num=", round(num, 6))
        print("den=", round(den, 6))
    print("J=", round(j, 6))
    print("crit=", crit)
    print("Reject restrictions" if j > crit else "Fail reject")
    print("Use larger model" if j > crit else "Restricted ok")


def jf(rssr, rssu, q, dfu, crit, mode=0):
    num = (rssr - rssu) / q
    den = rssu / dfu
    f = num / den
    if mode:
        print("F=((RSSr-RSSu)/q)/(RSSu/dfu)")
        print("num=", round(num, 6))
        print("den=", round(den, 6))
    print("F=", round(f, 6))
    print("crit=", crit)
    print("Reject restrictions" if f > crit else "Fail reject")
    print("Use larger model" if f > crit else "Restricted ok")


def adf(stat, crit, mode=0):
    if mode:
        print("ADF reject if stat < crit")
        print("stat=", stat)
        print("crit=", crit)
    if stat < crit:
        print("Reject unit root")
        print("Evidence stationary")
    else:
        print("Fail reject unit root")
        print("Treat non-stationary")


def ar1m(c, phi, s2, mode=0):
    m = c / (1 - phi)
    v = s2 / (1 - phi ** 2)
    if mode:
        print("mean=c/(1-phi)")
        print("var=s2/(1-phi^2)")
    print("mean=", round(m, 6))
    print("var=", round(v, 6))
    print("abs(phi)<1: stat ok" if abs(phi) < 1 else "abs(phi)>=1: not stationary")


def ar1f(c, phi, last, mode=0):
    mu = c + phi * last
    if mode:
        print("forecast=c+phi*last")
    print("forecast=", round(mu, 6))


def ar1fm(mu0, phi, last, mode=0):
    mu = mu0 + phi * (last - mu0)
    if mode:
        print("forecast=mu0+phi*(last-mu0)")
    print("forecast=", round(mu, 6))


def gar1(w, a, b, u, v, mode=0):
    shock = a * u ** 2
    persist = b * v
    vn = w + shock + persist
    sd = sqrt(vn)
    if mode:
        print("var=w+a*u^2+b*v")
        print("omega=", w)
        print("shock=", round(shock, 6))
        print("persist=", round(persist, 6))
    print("next var=", round(vn, 6))
    print("next sd=", round(sd, 6))


def gper(a, b, mode=0):
    p = a + b
    if mode:
        print("persist=alpha+beta")
    print("persist=", round(p, 6))
    if p < 1:
        print("Finite var condition ok")
        if p > 0.9:
            print("Highly persistent")
    else:
        print("Condition fails")


def pi1(mu, v, z, mode=0):
    sd = sqrt(v)
    lo = mu - z * sd
    hi = mu + z * sd
    if mode:
        print("PI=mu +/- z*sd")
        print("sd=", round(sd, 6))
    print("lower=", round(lo, 6))
    print("upper=", round(hi, 6))


def varp(value, mu, v, z, mode=0):
    sd = sqrt(v)
    q = mu - z * sd
    dv = abs(value * q / 100)
    if mode:
        print("Percent return VaR")
        print("sd=", round(sd, 6))
        print("q=mu-z*sd")
        print("VaR=abs(value*q/100)")
    print("q percent=", round(q, 6))
    print("VaR dollars=", round(dv, 2))
    print("Tail chance worse")


def vard(value, mu, v, z, mode=0):
    sd = sqrt(v)
    q = mu - z * sd
    dv = abs(value * q)
    if mode:
        print("Decimal return VaR")
        print("sd=", round(sd, 6))
        print("q=mu-z*sd")
        print("VaR=abs(value*q)")
    print("q decimal=", round(q, 6))
    print("VaR dollars=", round(dv, 2))
    print("Tail chance worse")


def arch(nobs, r2, q, crit, mode=0):
    stat = nobs * r2
    if mode:
        print("ARCH LM=T*R2")
        print("df=", q)
    print("LM=", round(stat, 6))
    print("crit=", crit)
    if stat > crit:
        print("Reject no ARCH")
        print("Vol time-varying")
    else:
        print("Fail reject")
        print("No ARCH found")


def emh(nobs, r2, p, crit, mode=0):
    stat = nobs * r2
    if mode:
        print("Residual AR test=T*R2")
        print("df=", p)
    print("stat=", round(stat, 6))
    print("crit=", crit)
    if stat > crit:
        print("Reject no autocorr")
        print("Against weak EMH")
    else:
        print("Fail reject")
        print("No EMH evidence against")


def jb(nobs, sk, ku, crit, mode=0):
    skpart = sk ** 2
    kpart = ((ku - 3) ** 2) / 4
    j = nobs / 6 * (skpart + kpart)
    if mode:
        print("JB=n/6*(S^2+(K-3)^2/4)")
        print("S^2=", round(skpart, 6))
        print("K part=", round(kpart, 6))
    print("JB=", round(j, 6))
    print("crit=", crit)
    if sk < 0:
        print("Skew<0: left tail")
    elif sk > 0:
        print("Skew>0: right tail")
    if ku > 3:
        print("K>3: fat tails")
    print("Reject normality" if j > crit else "Fail reject normal")


def gjr(a, lmb, mode=0):
    good = a
    bad = a + lmb
    if mode:
        print("good curvature=alpha")
        print("bad curvature=alpha+lambda")
    print("good=", round(good, 6))
    print("bad=", round(bad, 6))
    if lmb > 0:
        print("Bad news bigger effect")
    elif lmb == 0:
        print("Symmetric effect")
    else:
        print("Bad news smaller effect")


def aic(a1, a2, mode=0):
    if mode:
        print("Lower AIC preferred")
        print("model1 AIC=", a1)
        print("model2 AIC=", a2)
    if a1 < a2:
        print("Choose model 1")
    elif a2 < a1:
        print("Choose model 2")
    else:
        print("AIC tie")


def menu():
    print("")
    print("ETC3460 Toolkit")
    print("1 Coeff test")
    print("2 Risk split")
    print("3 GMV weights")
    print("4 Two-asset portfolio")
    print("5 RSS from RSE")
    print("6 Joint chi test")
    print("7 Joint F test")
    print("8 ADF decision")
    print("9 AR1 moments")
    print("10 AR1 forecast")
    print("11 Mean-adjusted AR forecast")
    print("12 GARCH forecast")
    print("13 GARCH persistence")
    print("14 Prediction interval")
    print("15 VaR percent returns")
    print("16 VaR decimal returns")
    print("17 ARCH LM")
    print("18 EMH residual test")
    print("19 Jarque-Bera")
    print("20 GJR curvature")
    print("21 AIC comparison")
    print("0 Exit")


def run_choice(ch):
    if ch == 1:
        print("type: 1 gen 2 beta 3 alpha 4 SMB 5 HML 6 MOM 7 lev")
        ctest(fnum("est: "), fnum("se: "), fnum("h0: "),
              fnum("crit: "), inum("type: "), mode_prompt())
    elif ch == 2:
        risk(fnum("R2: "), mode_prompt())
    elif ch == 3:
        gmv2(fnum("var1: "), fnum("var2: "), fnum("cov12: "), mode_prompt())
    elif ch == 4:
        port2(fnum("w1: "), fnum("ret1: "), fnum("ret2: "),
              fnum("var1: "), fnum("var2: "), fnum("cov12: "), mode_prompt())
    elif ch == 5:
        rssp(fnum("RSE: "), fnum("df: "), mode_prompt())
    elif ch == 6:
        jchi(fnum("RSS restricted: "), fnum("RSS unrestricted: "),
             fnum("df unrestricted: "), fnum("crit: "), mode_prompt())
    elif ch == 7:
        jf(fnum("RSS restricted: "), fnum("RSS unrestricted: "),
           fnum("q restrictions: "), fnum("df unrestricted: "), fnum("crit: "),
           mode_prompt())
    elif ch == 8:
        adf(fnum("ADF stat: "), fnum("crit: "), mode_prompt())
    elif ch == 9:
        ar1m(fnum("c: "), fnum("phi: "), fnum("innovation variance: "), mode_prompt())
    elif ch == 10:
        ar1f(fnum("c: "), fnum("phi: "), fnum("last: "), mode_prompt())
    elif ch == 11:
        ar1fm(fnum("mu0: "), fnum("phi: "), fnum("last: "), mode_prompt())
    elif ch == 12:
        gar1(fnum("omega: "), fnum("alpha: "), fnum("beta: "),
             fnum("last residual: "), fnum("last variance: "), mode_prompt())
    elif ch == 13:
        gper(fnum("alpha: "), fnum("beta: "), mode_prompt())
    elif ch == 14:
        pi1(fnum("mu: "), fnum("variance: "), fnum("z: "), mode_prompt())
    elif ch == 15:
        varp(fnum("position value: "), fnum("mu percent: "),
             fnum("variance: "), fnum("z: "), mode_prompt())
    elif ch == 16:
        vard(fnum("position value: "), fnum("mu decimal: "),
             fnum("variance: "), fnum("z: "), mode_prompt())
    elif ch == 17:
        arch(fnum("n/T: "), fnum("R2: "), fnum("q lags: "),
             fnum("crit: "), mode_prompt())
    elif ch == 18:
        emh(fnum("n/T: "), fnum("R2: "), fnum("p lags: "),
            fnum("crit: "), mode_prompt())
    elif ch == 19:
        jb(fnum("n: "), fnum("skew: "), fnum("kurtosis: "),
           fnum("crit: "), mode_prompt())
    elif ch == 20:
        gjr(fnum("alpha: "), fnum("lambda: "), mode_prompt())
    elif ch == 21:
        aic(fnum("AIC model 1: "), fnum("AIC model 2: "), mode_prompt())
    else:
        print("Unknown choice.")


def main():
    while True:
        menu()
        ch = inum("Choice: ")
        if ch == 0:
            print("Done.")
            break
        print("")
        run_choice(ch)
        print("")


main()
